src/googleclouddebugger/firebase_client.py

```diff
@@ -0,0 +1,571 @@
+
+    If the registration succeeds, sets self._debuggee_id to the registered
+    debuggee ID.
+
+    Args:
+      service: client to use for API calls
+
+    Returns:
+      (registration_required, delay) tuple
+    """
+    try:
+      debuggee = self._GetDebuggee()
+      self._debuggee_id = debuggee['id']
+
+      try:
+        debuggee_path = f'cdbg/debuggees/{self._debuggee_id}'
+        firebase_admin.db.reference(debuggee_path).set(debuggee)
+        native.LogInfo(
+            f'registering at {self._database_url}, path: {debuggee_path}')
+
+        native.LogInfo(
+            f'Debuggee registered successfully, ID: {self._debuggee_id}')
+        self.register_backoff.Succeeded()
+        return (False, 0)  # Proceed immediately to list active breakpoints.
+      except BaseException:
+        native.LogInfo(f'Failed to register debuggee: {traceback.format_exc()}')
+    except BaseException:
+      native.LogWarning('Debuggee information not available: ' +
+                        traceback.format_exc())
+
+    return (True, self.register_backoff.Failed())
+
+  def _SubscribeToBreakpoints(self):
+    # Kill any previous subscriptions first.
+    if self._breakpoint_subscription is not None:
+      self._breakpoint_subscription.close()
+      self._breakpoint_subscription = None
+
+    path = f'cdbg/breakpoints/{self._debuggee_id}/active'
+    native.LogInfo(f'Subscribing to breakpoint updates at {path}')
+    ref = firebase_admin.db.reference(path)
+    self._breakpoint_subscription = ref.listen(self._ActiveBreakpointCallback)
+
+  def _ActiveBreakpointCallback(self, event):
+    if event.event_type == 'put':
+      # Either a delete or a completely new set of breakpoints.
+      if event.data is None:
+        # Either deleting a breakpoint or initializing with no breakpoints.
+        # Initializing with no breakpoints is a no-op.
+        # If deleting, event.path will be /{breakpointid}
+        if event.path != '/':
+          breakpoint_id = event.path[1:]
+          del self._breakpoints[breakpoint_id]
+      else:
+        if event.path == '/':
+          # New set of breakpoints.
+          self._breakpoints = {}
+          for (key, value) in event.data.items():
+            self._AddBreakpoint(key, value)
+        else:
+          breakpoint_id = event.path[1:]
+          self._AddBreakpoint(breakpoint_id, breakpoint)
+
+    elif event.event_type == 'patch':
+      # New breakpoint or breakpoints.
+      for (key, value) in event.data.items():
+        self._AddBreakpoint(key, value)
+    else:
+      native.LogWarning('Unexpected event from Firebase: '
+                        f'{event.event_type} {event.path} {event.data}')
+      return
+
+    native.LogInfo(f'Breakpoints list changed, {len(self._breakpoints)} active')
+    self.on_active_breakpoints_changed(list(self._breakpoints.values()))
+
+  def _AddBreakpoint(self, breakpoint_id, breakpoint_data):
+    breakpoint_data['id'] = breakpoint_id
```

Comment:
In theory the CLI should ensure the ID is already part of the data, but not a bad idea to ensure it here.

Reply:
Yeah, the thought here is that the storage data model does not ensure that the breakpoint id is set, and the consequences of a missing breakpoint id elsewhere in the Python code are significant, so this is a little touch of defence in depth.